### ISSUE TITLE
Re-establish MongoDB support

### ISSUE THREAD

**Issue by gzoller** [gzoller/ScalaJack#87]

**Comment by tksmrch** [comment]
Hi, 

Does ScalaJack v5 support MongoDB? As I'm trying to upgrade to v5.0.7 from v4.8.3, I'm seeing that scalajack.rennder(obj) returns a BsonValue as opposed to a Document. Similarly, scalajack.read[T](v) requires a BsonValue as opposed to a Document. 

I really liked the feature in v4.8.3 that we could convert between Scala classes and Documents and store/read data from MongoDB seamlessly. As I've come to rely on this feature, I'm hoping that you've maintained it in v5 as well. 

Thank you for your awesome work and let me know if I'm missing anything.

**Comment by gzoller** [comment]
Hmm.  I remember making the change from Document to BsonValue (BsonDocument is also a BsonValue) but don't recall the reasoning.  That was also some time ago so possibly a different driver version.  I can take a look with the latest driver and play with going back to Document and see if the reason emerges, and if not cot 5.0.8 with Document support if that's better.

From your usage, what are pros and cons of each?  (I'm not a Mongo power user)

**Comment by gzoller** [comment]
Hi, tksmrch...  I've push a feature/mongo branch to this repo.  It passes unit tests and *seems* to work, but before I publish it, would you be able to build it locally and try it with some sample MongoDB use cases of yours and confirm this is what you had in mind?

Basically this just converted the BsonDocument to/from Document.  One note (added to docs) is that only the outermost object is a Document.  Any objects that are members of your outer object are still rendered as BsonDocuments.   This should be ok, but that's why it'd be great to see if this worked for you.

Let me know.
Thanks!

**Comment by tksmrch** [comment]
Hi Greg, 

Many thanks for the work on this... For my use cases, I simply need to convert my Scala (case) classes to/from Document in some way or the other, since Mongo Scala driver does not support BsonDocument. I can actually convert a BsonValue by 
`val doc = new Document(bsonValue.asDocument())`
, but I can NOT convert the other way round because 
`doc.underlying` is "protected."

It sounds like your remedy would work indeed. Please allow me a few days to test it on my end.  

cheers.

**Comment by gzoller** [comment]
Bump.  tksmrch, did you get a chance to look at this?

**Comment by tksmrch** [comment]
Greg,

Sorry about the delay. I'll get back to you tomorrow.

**Comment by tksmrch** [comment]
Hi Greg, 

I've tested the code and it seems there's something that needs fixing. See my test code and the output below. The same code works with ScalaJack 4.8.3. 

Thank you again for your help, 

### DataObjects.scala: 
```
package co.blocke.scalajack.mongo.test.data

import co.blocke.scalajack.DBKey
import scala.collection.immutable

case class Person(name: String, age: Int, isMale:Boolean)

case class Couple(husband: Person, wife:Person) {
  lazy val both = Seq(husband, wife)
}

case class Family(parents : Couple, children : immutable.Seq[Person],
                  assets : immutable.Map[String, Double]=Map.empty[String, Double]) {
  lazy val numMale = parents.both.count(_.isMale) + children.count(_.isMale)
}
```

### TestMongoDB.scala:
```
package co.blocke.scalajack.mongo.test.db

import java.util

import com.mongodb.ServerAddress
import org.mongodb.scala.connection.ClusterSettings
import org.mongodb.scala.{MongoClient, MongoClientSettings, MongoCredential}

import scala.collection.JavaConverters._

object TestMongoDB
{
  case class MongoConfig(user:String, password:String, dbCredential:String, host:String, port:Int, database:String)

  val DefaultConfig = MongoConfig(
    user           = "xxx",
    password       = "yyy",
    dbCredential   = "admin",
    host           = "zzz",
    port           = 26017,
    database       = "kkk"
  )

  lazy val mongoConf = DefaultConfig

  private val settings = {
    val credential      = MongoCredential.createCredential(
      mongoConf.user, mongoConf.dbCredential, mongoConf.password.toCharArray)

    val clusterSettings = ClusterSettings.builder().hosts(
      List(new ServerAddress(mongoConf.host, mongoConf.port)).asJava).build()

    MongoClientSettings.builder()
      .credentialList(util.Arrays.asList(credential))
      .clusterSettings(clusterSettings)
      .build()
  }

  lazy val Client   = MongoClient(settings)
  lazy val Database = Client.getDatabase(mongoConf.database)
}
```


### Test App:
```
package co.blocke.scalajack.mongo.test.app

import co.blocke.scalajack.ScalaJack
import co.blocke.scalajack.mongo.MongoFlavor
import co.blocke.scalajack.mongo.test.data.{Couple, Family, Person}
import co.blocke.scalajack.mongo.test.db.TestMongoDB

import scala.collection.immutable
import scala.concurrent.Await
import scala.concurrent.duration.Duration

object ScalaJackMongo extends App
{
  val database = TestMongoDB.Database

  val collectionName = "TestScalaJackMongo2"
  val collection = database.getCollection(collectionName)

  collection.drop()

  val p1 = Person("Bob", 35, true)
  val p2 = Person("Lynda", 38, false)
  val p3 = Person("David", 8, true)
  val p4 = Person("Leoni", 5, false)
  val p5 = Person("Susie", 3, false)
  
  val children = immutable.Seq(p3, p4, p5)
  val assets = immutable.Map("savings" -> 100.0, "morgage" -> 0d, "home" -> 200d)

  val c1 = Couple(p1, p2)
  val f1 = Family(c1, children, assets)

  val sj = ScalaJack(MongoFlavor())

  // Serialize to Document => WORKS
  val js = sj.render( f1)  
  println(js)

  // Insert to MongoDB => WORKS
  Await.result(collection.insertOne(js).toFuture(), Duration.Inf)

  // Retrieve Document from MongoDB => WORKS 
  val doc = Await.result(collection.find.first().toFuture(), Duration.Inf).head
  println(doc.toJson())

  // Convert Document back to object => ERROR
  val myObj = sj.read[Family](doc) // deserialization
  println(myObj)
}
```

### Output: 
```
Document((parents,{ "husband" : { "name" : "Bob", "age" : 35, "isMale" : true }, "wife" : { "name" : "Lynda", "age" : 38, "isMale" : false } }), (children,BsonArray{values=[{ "name" : "David", "age" : 8, "isMale" : true }, { "name" : "Leoni", "age" : 5, "isMale" : false }, { "name" : "Susie", "age" : 3, "isMale" : false }]}), (assets,{ "savings" : 100.0, "morgage" : 0.0, "home" : 200.0 }))
[Thread-7] INFO org.mongodb.driver.connection - Opened connection [connectionId{localValue:2, serverValue:935}] to scinfotrdb01:26017
{ "_id" : { "$oid" : "59976fbd41c3f820d5f50525" }, "parents" : { "husband" : { "name" : "Takeshi", "age" : 35, "isMale" : true }, "wife" : { "name" : "Manami", "age" : 38, "isMale" : false } }, "children" : [{ "name" : "Shunri", "age" : 8, "isMale" : true }, { "name" : "Yui", "age" : 5, "isMale" : false }, { "name" : "Satsuki", "age" : 3, "isMale" : false }], "assets" : { "savings" : 100.0, "morgage" : 0.0, "home" : 200.0 } }
Exception in thread "main" java.lang.NumberFormatException: For input string: "TOKEN TEXT"
SOMETHING WENT WRONG
	at co.blocke.scalajack.Reader$class.readDouble(Reader.scala:116)
	at co.blocke.scalajack.mongo.BsonReader.readDouble(BsonReader.scala:7)
	at co.blocke.scalajack.typeadapter.DoubleTypeAdapter$.read(DoubleTypeAdapter.scala:7)
	at co.blocke.scalajack.typeadapter.DoubleTypeAdapter$.read(DoubleTypeAdapter.scala:4)
	at co.blocke.scalajack.typeadapter.CanBuildMapTypeAdapter.read(CanBuildFromTypeAdapter.scala:95)
	at co.blocke.scalajack.typeadapter.CanBuildMapTypeAdapter.read(CanBuildFromTypeAdapter.scala:80)
	at co.blocke.scalajack.typeadapter.CaseClassTypeAdapter$FieldMember.readValue(CaseClassTypeAdapter.scala:74)
	at co.blocke.scalajack.typeadapter.CaseClassTypeAdapter.read(CaseClassTypeAdapter.scala:274)
	at co.blocke.scalajack.mongo.MongoFlavor.read(MongoFlavor.scala:39)
	at co.blocke.scalajack.mongo.MongoFlavor.read(MongoFlavor.scala:11)
	at co.blocke.scalajack.mongo.test.app.ScalaJackMongo$.delayedEndpoint$co$blocke$scalajack$mongo$test$app$ScalaJackMongo$1(ScalaJackMongo.scala:55)
	at co.blocke.scalajack.mongo.test.app.ScalaJackMongo$delayedInit$body.apply(ScalaJackMongo.scala:19)
	at scala.Function0$class.apply$mcV$sp(Function0.scala:34)
	at scala.runtime.AbstractFunction0.apply$mcV$sp(AbstractFunction0.scala:12)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.App$$anonfun$main$1.apply(App.scala:76)
	at scala.collection.immutable.List.foreach(List.scala:381)
	at scala.collection.generic.TraversableForwarder$class.foreach(TraversableForwarder.scala:35)
	at scala.App$class.main(App.scala:76)
	at co.blocke.scalajack.mongo.test.app.ScalaJackMongo$.main(ScalaJackMongo.scala:19)
	at co.blocke.scalajack.mongo.test.app.ScalaJackMongo.main(ScalaJackMongo.scala)
```

**Comment by gzoller** [comment]
Haven't forgotten about you!  Had some business travel that distracted me a bit.  I've been able to confirm this error with your given example, so that's an important start.  I was also able to insert/read (successfully) a simple Person object, so the problem is involves these more complex objects.  Will be digging in, and will let you know when I learn more about what's happening.

**Comment by tksmrch** [comment]
Thank you, Greg - I'll be happy to help you test the code during the dev
process.  Let me know.

**Comment by gzoller** [comment]
Ok, I've made a fix on the mongo feature branch.  It works in my tests.  It was not handling Double properly.  Please have a peek and let me know if it works for you as well.  Thanks!

**Comment by tksmrch** [comment]
Thank you for your work! I've confirmed it's working on my end. 
Would you please release this as 5.0.8?

**Comment by gzoller** [comment]
Published 5.0.8
